### PR TITLE
Get GPU topo information from topo cache

### LINKED ISSUE #2800: GEOPM Topo cache file should contain GPU information
**Is your feature request related to a problem? Please describe.**
I_m always frustrated when I try to find the service topology by checking the geopm topo cache and it only contains CPU information

**Describe the solution you'd like**
GPU information should be provided in the topo cache.  

**Describe alternatives you've considered**
Rely on the user ability to access the GPU Topo instead of the service's cached topo.

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Fixes #2800 change request from github issues.

### CODE REVIEW

File: service/src/PlatformTopo.cpp [hunk 1]
```diff
@@ -525,7 +495,30 @@ namespace geopm
                 throw Exception("PlatformTopo::create_cache(): Could not pclose lscpu command: ",
                                 errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
             }
-
+            const GPUTopo &gtopo = gpu_topo();
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]:
```suggestion
            const GPUTopo &gtopo = geopm::gpu_topo();
```

Comment (bgeltz) [inline service/src/PlatformTopo.cpp]: Somewhere in here there should be a throw to see if ZE_AFFINITY_MASK is set.  It's unfortunate that it's a L0 specific env var.  Not sure if NVML has a similar feature.

File: service/src/PlatformTopo.cpp [hunk 2]
```diff
@@ -525,7 +495,30 @@ namespace geopm
...
+            if (gtopo.num_gpu() != 0) {
+                std::ofstream cache_stream;
+                cache_stream.open(tmp_path, std::ios_base::app);
+                std::vector<int> gpu_domains = {
+                    GEOPM_DOMAIN_GPU,
+                    GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU,
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]:
```suggestion
```

File: service/src/PlatformTopo.cpp [hunk 3]
```diff
@@ -630,6 +623,31 @@ namespace geopm
                 numa_map[0].insert(cpu_idx);
             }
         }
+        return numa_map;
+    }
+
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp:627]:
```suggestion
```

File: service/src/PlatformTopo.cpp [hunk 4]
```diff
@@ -115,23 +116,20 @@ namespace geopm
...
+              {GEOPM_DOMAIN_GPU_CHIP, {}}
+          }
     {
         std::map<std::string, std::string> lscpu_map;
         lscpu(lscpu_map);
         parse_lscpu(lscpu_map, m_num_package, m_core_per_package, m_thread_per_core);
-        parse_lscpu_numa(lscpu_map, m_numa_map);
+        m_numa_map = parse_lscpu_numa(lscpu_map);
+        m_gpu_info[GEOPM_DOMAIN_GPU] = parse_lscpu_gpu(lscpu_map, GEOPM_DOMAIN_GPU);
+        m_gpu_info[GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU] = parse_lscpu_gpu(lscpu_map, GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU);
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]:
```suggestion
```

File: service/src/PlatformTopo.cpp [hunk 5]
```diff
@@ -115,23 +116,20 @@ namespace geopm
...
-        parse_lscpu_numa(lscpu_map, m_numa_map);
-    }
-
-    PlatformTopoImp::PlatformTopoImp(const std::string &test_cache_file_name,
-                                     const GPUTopo &gpu_topo)
         : M_TEST_CACHE_FILE_NAME(test_cache_file_name)
-        , m_gpu_topo(gpu_topo)
+        , m_gpu_info {
+              {GEOPM_DOMAIN_GPU, {}},
+              {GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU, {}},
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]:
```suggestion
```

File: service/src/PlatformTopo.cpp [hunk 6]
```diff
@@ -630,6 +623,31 @@ namespace geopm
...
+            gpu_key << "GPU " << short_name << domain_idx << " CPU(s)";
+            auto lscpu_it = lscpu_map.find(gpu_key.str());
+            if (lscpu_it == lscpu_map.end()) {
+                is_domain_found = false;
+            }
+            else {
+                gpu_map.push_back({});
+                auto cpu_set_it = gpu_map.end() - 1;
+                for (const auto &cpu_str : string_split(lscpu_it->second, ",")) {
+                    cpu_set_it->insert(stoi(cpu_str));
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]:
```suggestion
                    cpu_set_it->insert(std::stoi(cpu_str));
```

File: service/src/PlatformTopo.cpp [hunk 7]
```diff
@@ -630,6 +623,31 @@ namespace geopm
...
+            std::ostringstream gpu_key;
+            gpu_key << "GPU " << short_name << domain_idx << " CPU(s)";
+            auto lscpu_it = lscpu_map.find(gpu_key.str());
+            if (lscpu_it == lscpu_map.end()) {
+                is_domain_found = false;
+            }
+            else {
+                gpu_map.push_back({});
+                auto cpu_set_it = gpu_map.end() - 1;
+                for (const auto &cpu_str : string_split(lscpu_it->second, ",")) {
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]:
```suggestion
                for (const auto &cpu_str : geopm::string_split(lscpu_it->second, ",")) {
```

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: service/src/PlatformTopo.cpp (hunk 1 above)
Comment (cmcantalupo) [reply service/src/PlatformTopo.cpp]: I just put the check in the constructor for the LevelZeroGPUTopo.

File: service/src/PlatformTopo.cpp [hunk 8]
```diff
@@ -431,10 +431,10 @@ namespace geopm
 
     void PlatformTopoImp::create_cache(void)
     {
-        try {
+        if (getuid() == 0) {
```
Comment (cmcantalupo) [inline service/src/PlatformTopo.cpp:430]: geteuid()?

File: service/src/PlatformTopo.cpp [hunk 9]
```diff
@@ -699,14 +699,14 @@ namespace geopm
             result = geopm::read_file(M_TEST_CACHE_FILE_NAME);
         }
         else {
-            try {
-                create_cache(M_CACHE_FILE_NAME);
-                result = geopm::read_file(M_CACHE_FILE_NAME);
-            }
-            catch (const geopm::Exception &ex) {
+            if (getuid() == 0) {
```
Comment (cmcantalupo) [inline service/src/PlatformTopo.cpp:701]: geteuid()?

File: service/src/PlatformTopo.cpp [hunk 10]
```diff
@@ -485,6 +456,7 @@ namespace geopm
         }
 
         if (is_file_ok == false) {
+
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]:
```suggestion
```

**Review by bgeltz — CHANGES_REQUESTED** [review]

**Review by bgeltz — APPROVED** [review]

**Pull request merged**